2022q1 release

rleh opened the pull request:
It's quarterly release time again! 2022-04-01 is the date.

- [x] Update all submodules
- [x] Fix submodule issues
  - [x] TinyUSB: Enable on STM32G4 (Feature!)
  - [x] TinyUSB: Fix hard_assert error with RP2040
  - [x] LVGL: `lv_indev.c:413:21: error: format '%d' expects argument of type 'int', but argument 6 has type 'uint32_t' {aka 'long unsigned int'} [-Werror=format=]`
- [x] Merge remaining PRs?
  - [x] #841
  - [ ] ~~#769~~
  - [ ] ~~#817~~
- [x] Deprecate all stuff marked for deprecation
  - [x] Remove usage of removed `modm::PeriodicTimer` constructor in `modm::BitBangEncoderOutput`
- [x] Rebuild docker images and use new images
- [x] Check for updated compilers
- [x] Update release notes
- [ ] Hardware testing
  - [x] Ethernet/FreeRTOS on Nucleo-F429ZI
  - [x] TinyUSB + FreeRTOS on Nucleo-F429ZI
  - [x] LVGL on DISCO-F469NI
  - [x] TinyUSB on RP2040 (Raspberry Pi Pico)
  - [x] Unittests on Linux
  - [ ] Unittests on MacOS
  - [ ] ~~Unittests on Windows~~
  - [x] Unittests on Nucleo-F091RC
  - [x] Unittests on Nucleo-F411RC
  - [x] Unittests on Nucleo-L552ZE
  - [x] Unittests on Nucleo-L432KC
  - [x] Unittests on Nucleo-F103RB
  - [ ] ~~Unittests on Arduino Uno~~
  - [ ] ~~Unittests on AVR Mega 2560 Pro~~
- [x] Fix author.py script to detect Co-authored-by correctly. ~~next time™~~

rleh committed:
[ext] Update submodules

rleh committed:
[ci] Update docker image to 2022-03-28

chris-durand committed:
[stm32] Use correct tinyusb driver for G4 USB

rleh committed:
Remove deprecated code

rleh committed:
RP Pico compatibility for tinyusb: hard_assert

chris-durand commented:
>  [x] TinyUSB: Fix hard_assert error with RP2040

How was that fixed? The compilation of RP2040 examples still fails. The CI seems to have swallowed the compile errors but there still is "exit code 1".

rleh commented:
Give me 1 second to push commit 3d8ee01 ;)

salkinium commented:
modm-devices also needs to be fixed, the CI is failing since a new STM32U5 was released and it's missing the memory model.

rleh committed:
BitBangEncoderOutput: Fix usage of PeriodicTimer contructor

chris-durand committed:
[examples] Fix outdated use of duration types

chris-durand committed:
[platform] Fix I2cMaster device reset parameter

rleh committed:
[tools] Improve authors.py script to recognize co-authors

rleh committed:
New contributor: Rasmus Kleist

rleh committed:
Add 2022q1 release notes

salkinium commented on docs/release/2022q1.md:
```diff
@@ -0,0 +1,144 @@
+
+Integrated Projects:
+
+- ETL added at v20.27.1.
```
```suggestion
- ETL upgraded to v20.27.1.
```

salkinium commented on docs/release/2022q1.md:
```diff
@@ -0,0 +1,144 @@
+
+- Support for newer revision (B-03) of the DISCO-F469NI development board
+  as `modm:disco-f469ni:b-03`
+- Nucleo-L552ZE-Q as `modm:board:nucleo-l552ze-q`
```
```suggestion
- Nucleo-L552ZE-Q as `modm:board:nucleo-l552ze-q`.
- RP-PICO as `modm:board:rp-pico`.
```

salkinium commented on docs/release/2022q1.md:
```diff
@@ -0,0 +1,144 @@
+New development boards:
+
+- Support for newer revision (B-03) of the DISCO-F469NI development board
+  as `modm:disco-f469ni:b-03`
```
```suggestion
  as `modm:disco-f469ni:b-03`.
```

salkinium commented on docs/release/2022q1.md:
```diff
@@ -0,0 +1,144 @@
+PR #800 -> 187ddd8.  
+Tested in hardware by @chris-durand.
+
+#### 2022-01-19: Changed tilde (`~`) to `$HOME` variable due to lack of expansion
```
This is just documentation?

chris-durand replied:
Yes, should not be in the change log.

salkinium commented on docs/release/2022q1.md:
```diff
@@ -0,0 +1,144 @@
+
+PR #813 -> 77a761d.
+
+#### 2022-01-07: Scons: Emit explicit error when chosen gcc isn't found on PATH
```
Kinda not relevant?

salkinium commented on docs/release/2022q1.md:
```diff
@@ -0,0 +1,144 @@
+development board.
+
+PR #806 -> 1c9c0b6.  
+Tested in hardware by @andryblack and @rleh.
```
```suggestion
Tested in hardware by @andryblack, @rleh and @salkinium.
```

salkinium commented on docs/release/2022q1.md:
```diff
@@ -0,0 +1,144 @@
+- Issues when using TinyUSB and FreeRTOS together solved.
+- Fix SPI bug in handling of CPHA/CPOL on Atmel SAM.
+- Read unique ID correctly on SAMG55.  
+- Fix GPIO peripheral connect method on Atmel SAM.
```
```suggestion
- Fix GPIO peripheral connect method on Atmel SAM.
- Heap allocator defaults to Newlib on devices with one continuous RAM section.
- FreeRTOS now always supports static allocation.
```

(tiny, but important change in fa44651)

chris-durand replied:
There's no `Atmel` anymore, I'll update it to `Microchip`.

salkinium commented on docs/release/2022q1.md:
```diff
@@ -0,0 +1,144 @@
+
+Fixes:
+
+- Issues when using TinyUSB and FreeRTOS together solved.
```
```suggestion
- Issues when using TinyUSB and FreeRTOS together solved.
- Fix TinyUSB crash when reading device descriptor.
- TinyUSB correctly chooses fullspeed port on STM32 by default.
```

salkinium commented on docs/release/2022q1.md:
```diff
@@ -0,0 +1,144 @@
+
+Breaking changes:
+
+None.
```
```suggestion
- Removed all functions marked deprecated for >1 year.
```

salkinium reviewed (commented):
Thanks for doing the release!

But…… I have opinions…

salkinium commented:
> modm-devices also needs to be fixed, the CI is failing since a new STM32U5 was released and it's missing the memory model.

I guess we'll do it next time?

salkinium commented on src/modm/platform/i2c/bitbang/bitbang_i2c_master.hpp.in:
```diff
@@ -51,9 +51,9 @@ class BitBangI2cMaster : public modm::I2cMaster
+	template< class... Signals >
 	static void
-	connect(PullUps pullups = PullUps::External);
+	connect(PullUps pullups = PullUps::External, ResetDevices reset = ResetDevices::Standard);
```
The interface also needs to be adapted:
https://github.com/modm-io/modm/blob/develop/src/modm/architecture/interface/i2c_master.hpp#L89-L91

chris-durand replied:
:+1:, I'll fix it.

chris-durand commented:
> modm-devices also needs to be fixed, the CI is failing since a new STM32U5 was released and it's missing the memory model.

No STM32U5 targets are enabled in modm yet. Before we do that I'd say it's not that critical for the release and we can move it.

salkinium commented:
Christopher should be added as maintainer to the docs too:
- https://github.com/modm-io/modm/blob/develop/README.md#how-can-i-contribute
- https://github.com/modm-io/modm/blob/develop/CONTRIBUTING.md#when-in-doubt-ask
- more?

chris-durand committed:
[docs] Add Christopher as maintainer

chris-durand commented:
> Christopher should be added as maintainer to the docs too

Done

rleh commented:
### Unittests Nucleo F103 A

<details>
<summary>Failed 18 of 2645 tests</summary>

```
Unittests (Apr  2 2022, 16:31:06)␊
Machine:  rleh-pc␊
User:     rleh␊
Os:       Linux-5.16.16-200.fc35.x86_64-x86_64-with-glibc2.34␊
Compiler: arm-none-eabi-g++ 10.2.1␊
Local Git User:␊
Name:  Raphael Lehmann␊
Email: raphael@rleh.de␊
Last Commit:␊
SHA:             175f915e272a1663025e36a515d3f135353cceae␊
Abbreviated SHA: 175f915e27␊
Subject:         [docs] Add Christopher as maintainer␊
␊
Author:␊
Name:      Christopher Durand␊
Email:     christopher.durand@rwth-aachen.de␊
Date:      2022-04-01 22:37:05+02:00␊
Timestamp: 1648845425␊
␊
Committer:␊
Name:      Christopher Durand␊
Email:     christopher.durand@rwth-aachen.de␊
Date:      2022-04-01 22:37:05+02:00␊
Timestamp: 1648845425␊
␊
File Status:␊
Modified:  0␊
Added:     0␊
Deleted:   0␊
Renamed:   0␊
Copied:    0␊
Untracked: 0␊
␊
>>> ad7280a_test␊
>>> atomic_queue_test␊
>>> block_allocator_test␊
>>> bme280_test␊
>>> bmp085_test␊
>>> bounded_deque_test␊
>>> bounded_queue_test␊
>>> bounded_stack_test␊
>>> button_group_test␊
>>> button_test␊
>>> can_bit_timings_test␊
>>> can_lawicel_formatter_test␊
>>> can_message_test␊
>>> clock_test␊
>>> color_test␊
>>> delay_test␊
>>> doubly_linked_list_test␊
>>> drv832x_spi_test␊
>>> dynamic_array_test␊
>>> fiber_test␊
>>> flash_test␊
>>> gpio_sampler_test␊
Channel 0:␊
-2002371004␊
-2002370859␊
-2002306698␊
-2002242529␊
-2002178364␊
Channel 1:␊
-2002371003␊
-2002114196␊
-2002050035␊
-2001985876␊
-2001921717␊
FAIL: gpio_sampler_test:71 : -145 not in range [0,64000]␊
FAIL: gpio_sampler_test:72 : -64161 not in range [64000,70400]␊
FAIL: gpio_sampler_test:73 : -64169 not in range [64000,70400]␊
FAIL: gpio_sampler_test:74 : -64165 not in range [64000,70400]␊
FAIL: gpio_sampler_test:77 : -256807 not in range [256000,262400]␊
FAIL: gpio_sampler_test:78 : -64161 not in range [64000,70400]␊
FAIL: gpio_sampler_test:79 : -64159 not in range [64000,70400]␊
FAIL: gpio_sampler_test:80 : -64159 not in range [64000,70400]␊
>>> i2c_test␊
>>> io_stream_test␊
>>> linked_list_test␊
>>> ltc2984_test␊
>>> mcp2515_can_bit_timings_test␊
>>> pair_test␊
>>> periodic_timer_test␊
>>> platform_gpio_test␊
FAIL: platform_gpio_test:65 : true == false␊
FAIL: platform_gpio_test:69 : true == false␊
FAIL: platform_gpio_test:70 : true == false␊
>>> protothread_test␊
>>> register_test␊
>>> resumable_test␊
>>> scheduler_test␊
>>> spi_bitbang_test␊
FAIL: spi_bitbang_test:38 : 1 == 17␊
failed at 'h' with '-1999037312'␊
failed at 'h' with '-1999037312'␊
failed at 'h' with '-1999037312'␊
failed at 'h' with '-1999037312'␊
failed at 'h' with '-1999037312'␊
failed at 'h' with '-1999037312'␊
failed at 'h' with '-1999037312'␊
failed at 'h' with '-1999037312'␊
  0 -1999037312      0 (0us)␊
␊
  0 -1999037312      0 (0us)␊
␊
FAIL: spi_bitbang_test:48 : 1 == 9␊
FAIL: spi_bitbang_test:49 : 1 == 17␊
failed at 'h' with '-1998911616'␊
failed at 'h' with '-1998911616'␊
failed at 'h' with '-1998911616'␊
failed at 'h' with '-1998911616'␊
failed at 'h' with '-1998911616'␊
failed at 'h' with '-1998911616'␊
failed at 'h' with '-1998911616'␊
failed at 'h' with '-1998911616'␊
failed at 'h' with '-1998911616'␊
failed at 'h' with '-1998911616'␊
failed at 'h' with '-1998911616'␊
failed at 'h' with '-1998911616'␊
  0 -1998911616      0 (0us)␊
␊
FAIL: spi_bitbang_test:58 : 1 == 3␊
FAIL: spi_bitbang_test:59 : 1 == 17␊
failed at 'h' with '-1998760344'␊
failed at 'h' with '-1998760344'␊
failed at 'h' with '-1998760344'␊
failed at 'h' with '-1998760344'␊
failed at 'h' with '-1998760344'␊
failed at 'h' with '-1998760344'␊
failed at 'h' with '-1998760344'␊
failed at 'h' with '-1998760344'␊
failed at 'h' with '-1998760344'␊
  0 -1998760344      0 (0us)␊
␊
FAIL: spi_bitbang_test:69 : 1 == 3␊
FAIL: spi_bitbang_test:70 : 1 == 17␊
  0 -1998628940      0 (0us)␊
␊
failed at 'h' with '-1998628940'␊
failed at 'h' with '-1998628940'␊
failed at 'h' with '-1998628940'␊
failed at 'h' with '-1998628940'␊
failed at 'h' with '-1998628940'␊
failed at 'h' with '-1998628940'␊
failed at 'h' with '-1998628940'␊
failed at 'h' with '-1998628940'␊
failed at 'h' with '-1998628940'␊
>>> spi_device_test␊
>>> time_test␊
>>> timeout_test␊
>>> timestamp_test␊
>>> tmp12x_test␊
␊
Failed 18 of 2645 tests␊
FAIL!␊
```

</details>


### Unittests Nucleo F103 B

<details>
<summary>Passed 2060 tests</summary>

```
<0xff>Unittests (Apr  2 2022, 16:32:44)␊
Machine:  rleh-pc␊
User:     rleh␊
Os:       Linux-5.16.16-200.fc35.x86_64-x86_64-with-glibc2.34␊
Compiler: arm-none-eabi-g++ 10.2.1␊
Local Git User:␊
Name:  Raphael Lehmann␊
Email: raphael@rleh.de␊
Last Commit:␊
SHA:             175f915e272a1663025e36a515d3f135353cceae␊
Abbreviated SHA: 175f915e27␊
Subject:         [docs] Add Christopher as maintainer␊
␊
Author:␊
Name:      Christopher Durand␊
Email:     christopher.durand@rwth-aachen.de␊
Date:      2022-04-01 22:37:05+02:00␊
Timestamp: 1648845425␊
␊
Committer:␊
Name:      Christopher Durand␊
Email:     christopher.durand@rwth-aachen.de␊
Date:      2022-04-01 22:37:05+02:00␊
Timestamp: 1648845425␊
␊
File Status:␊
Modified:  0␊
Added:     0␊
Deleted:   0␊
Renamed:   0␊
Copied:    0␊
Untracked: 0␊
␊
>>> LUDecomposition_test␊
>>> angle_test␊
>>> arithmetic_traits_test␊
>>> bit_operation_test␊
>>> circle_2d_test␊
>>> debounce_test␊
>>> endianness_test␊
>>> fir_test␊
>>> lagrange_interpolation_test␊
>>> line_2d_test␊
>>> line_segment_2d_test␊
>>> linear_interpolation_test␊
>>> location_2d_test␊
>>> matrix_test␊
>>> matrix_vector_test␊
>>> median_test␊
>>> moving_average_test␊
>>> operator_test␊
>>> pid_test␊
>>> point_set_2d_test␊
>>> polygon_2d_test␊
>>> prescaler_test␊
>>> quaternion_test␊
>>> ramp_test␊
>>> range_test␊
>>> ray_2d_test␊
>>> s_curve_controller_test␊
>>> s_curve_generator_test␊
>>> saturation_test␊
>>> vector1_test␊
>>> vector2_test␊
>>> vector3_test␊
>>> vector4_test␊
>>> vector_test␊
␊
Passed 2060 tests␊
OK!␊
```

</details>


### Unittests Nucleo L432

<details>
<summary>Passed 4876 tests</summary>

```
Unittests (Apr  2 2022, 16:33:42)␊
Machine:  rleh-pc␊
User:     rleh␊
Os:       Linux-5.16.16-200.fc35.x86_64-x86_64-with-glibc2.34␊
Compiler: arm-none-eabi-g++ 10.2.1␊
Local Git User:␊
Name:  Raphael Lehmann␊
Email: raphael@rleh.de␊
Last Commit:␊
SHA:             175f915e272a1663025e36a515d3f135353cceae␊
Abbreviated SHA: 175f915e27␊
Subject:         [docs] Add Christopher as maintainer␊
␊
Author:␊
Name:      Christopher Durand␊
Email:     christopher.durand@rwth-aachen.de␊
Date:      2022-04-01 22:37:05+02:00␊
Timestamp: 1648845425␊
␊
Committer:␊
Name:      Christopher Durand␊
Email:     christopher.durand@rwth-aachen.de␊
Date:      2022-04-01 22:37:05+02:00␊
Timestamp: 1648845425␊
␊
File Status:␊
Modified:  0␊
Added:     0␊
Deleted:   0␊
Renamed:   0␊
Copied:    0␊
Untracked: 0␊
␊
>>> LUDecomposition_test␊
>>> ad7280a_test␊
>>> angle_test␊
>>> arithmetic_traits_test␊
>>> atomic_queue_test␊
>>> bit_operation_test␊
>>> block_allocator_test␊
>>> bme280_test␊
>>> bmp085_test␊
>>> bounded_deque_test␊
>>> bounded_queue_test␊
>>> bounded_stack_test␊
>>> button_group_test␊
>>> button_test␊
>>> can_bit_timings_test␊
>>> can_connector_base_test␊
>>> can_connector_test␊
>>> can_lawicel_formatter_test␊
>>> can_message_test␊
>>> circle_2d_test␊
>>> clock_test␊
>>> color_test␊
>>> debounce_test␊
>>> delay_test␊
>>> dispatcher_test␊
>>> doubly_linked_list_test␊
>>> drv832x_spi_test␊
>>> dynamic_array_test␊
>>> endianness_test␊
>>> fiber_test␊
>>> fir_test␊
>>> flash_test␊
>>> i2c_test␊
>>> inplace_any_test␊
>>> interface_test␊
>>> interface_test␊
>>> io_stream_test␊
>>> lagrange_interpolation_test␊
>>> line_2d_test␊
>>> line_segment_2d_test␊
>>> linear_interpolation_test␊
>>> linked_list_test␊
>>> location_2d_test␊
>>> ltc2984_test␊
>>> matrix_test␊
>>> matrix_vector_test␊
>>> mcp2515_can_bit_timings_test␊
>>> median_test␊
>>> message_test␊
>>> moving_average_test␊
>>> node_test␊
>>> operator_test␊
>>> pair_test␊
>>> periodic_timer_test␊
>>> pid_test␊
>>> point_set_2d_test␊
>>> polygon_2d_test␊
>>> prescaler_test␊
>>> protothread_test␊
>>> quaternion_test␊
>>> ramp_test␊
>>> range_test␊
>>> ray_2d_test␊
>>> register_test␊
>>> resumable_test␊
>>> s_curve_controller_test␊
>>> s_curve_generator_test␊
>>> saturation_test␊
>>> scheduler_test␊
>>> slave_test␊
>>> spi_device_test␊
>>> time_test␊
>>> timeout_test␊
>>> timestamp_test␊
>>> tmp12x_test␊
>>> vector1_test␊
>>> vector2_test␊
>>> vector3_test␊
>>> vector4_test␊
>>> vector_test␊
␊
Passed 4876 tests␊
OK!␊
```

</details>


### Unittests Nucleo F411

<details>
<summary>Failed 11 of 6026 tests</summary>

```
<break>
Unittests (Apr  2 2022, 16:33:17)␊
Machine:  rleh-pc␊
User:     rleh␊
Os:       Linux-5.16.16-200.fc35.x86_64-x86_64-with-glibc2.34␊
Compiler: arm-none-eabi-g++ 10.2.1␊
Local Git User:␊
Name:  Raphael Lehmann␊
Email: raphael@rleh.de␊
Last Commit:␊
SHA:             175f915e272a1663025e36a515d3f135353cceae␊
Abbreviated SHA: 175f915e27␊
Subject:         [docs] Add Christopher as maintainer␊
␊
Author:␊
Name:      Christopher Durand␊
Email:     christopher.durand@rwth-aachen.de␊
Date:      2022-04-01 22:37:05+02:00␊
Timestamp: 1648845425␊
␊
Committer:␊
Name:      Christopher Durand␊
Email:     christopher.durand@rwth-aachen.de␊
Date:      2022-04-01 22:37:05+02:00␊
Timestamp: 1648845425␊
␊
File Status:␊
Modified:  0␊
Added:     0␊
Deleted:   0␊
Renamed:   0␊
Copied:    0␊
Untracked: 0␊
␊
>>> LUDecomposition_test␊
>>> ad7280a_test␊
>>> angle_test␊
>>> arithmetic_traits_test␊
>>> atomic_queue_test␊
>>> bit_operation_test␊
>>> block_allocator_test␊
>>> bme280_test␊
>>> bmp085_test␊
>>> bounded_deque_test␊
>>> bounded_queue_test␊
>>> bounded_stack_test␊
>>> button_group_test␊
>>> button_test␊
>>> can_bit_timings_test␊
>>> can_connector_base_test␊
>>> can_connector_test␊
>>> can_lawicel_formatter_test␊
>>> can_message_test␊
>>> circle_2d_test␊
>>> clock_test␊
>>> color_test␊
>>> debounce_test␊
>>> delay_test␊
>>> dispatcher_test␊
>>> doubly_linked_list_test␊
>>> drv832x_spi_test␊
>>> dynamic_array_test␊
>>> endianness_test␊
>>> fiber_test␊
>>> fir_test␊
>>> flash_test␊
>>> gpio_sampler_test␊
Channel 0:␊
1673985936␊
1673986071␊
1674082196␊
1674178309␊
1674274424␊
Channel 1:␊
1673985937␊
1674370546␊
1674466663␊
1674562832␊
1674658945␊
FAIL: gpio_sampler_test:71 : -135 not in range [0,96000]␊
FAIL: gpio_sampler_test:72 : -96125 not in range [96000,105600]␊
FAIL: gpio_sampler_test:73 : -96113 not in range [96000,105600]␊
FAIL: gpio_sampler_test:74 : -96115 not in range [96000,105600]␊
FAIL: gpio_sampler_test:77 : -384609 not in range [384000,393600]␊
FAIL: gpio_sampler_test:78 : -96117 not in range [96000,105600]␊
FAIL: gpio_sampler_test:79 : -96169 not in range [96000,105600]␊
FAIL: gpio_sampler_test:80 : -96113 not in range [96000,105600]␊
>>> i2c_test␊
>>> inplace_any_test␊
>>> interface_test␊
>>> interface_test␊
>>> io_stream_test␊
>>> lagrange_interpolation_test␊
>>> line_2d_test␊
>>> line_segment_2d_test␊
>>> linear_interpolation_test␊
>>> linked_list_test␊
>>> location_2d_test␊
>>> ltc2984_test␊
>>> matrix_test␊
>>> matrix_vector_test␊
>>> mcp2515_can_bit_timings_test␊
>>> median_test␊
>>> message_test␊
>>> moving_average_test␊
>>> node_test␊
>>> operator_test␊
>>> pair_test␊
>>> periodic_timer_test␊
>>> pid_test␊
>>> platform_gpio_test␊
FAIL: platform_gpio_test:66 : true == false␊
FAIL: platform_gpio_test:74 : true == false␊
FAIL: platform_gpio_test:79 : true == false␊
>>> point_set_2d_test␊
>>> polygon_2d_test␊
>>> prescaler_test␊
>>> protothread_test␊
>>> quaternion_test␊
>>> ramp_test␊
>>> range_test␊
>>> ray_2d_test␊
>>> register_test␊
>>> resumable_test␊
>>> s_curve_controller_test␊
>>> s_curve_generator_test␊
>>> saturation_test␊
>>> scheduler_test␊
>>> slave_test␊
>>> spi_bitbang_test␊
failed at 'l' with '1712766950'␊
failed at 'l' with '1712766950'␊
failed at 'h' with '1712772009'␊
failed at 'l' with '1712776946' at diff -4937 '-4320'␊
failed at 'l' with '1712776946'␊
failed at 'h' with '1712781833' at diff -4887 '-4320'␊
failed at 'l' with '1712786762' at diff -4929 '-4320'␊
failed at 'l' with '1712786762'␊
failed at 'h' with '1712791649' at diff -4887 '-4320'␊
failed at 'l' with '1712796582' at diff -4933 '-4320'␊
failed at 'l' with '1712796582'␊
failed at 'h' with '1712801539' at diff -4957 '-4320'␊
failed at 'l' with '1712806402' at diff -4863 '-4320'␊
failed at 'l' with '1712806402'␊
failed at 'h' with '1712811359' at diff -4957 '-4320'␊
failed at 'l' with '1712816222' at diff -4863 '-4320'␊
failed at 'l' with '1712816222'␊
failed at 'h' with '1712821179' at diff -4957 '-4320'␊
failed at 'l' with '1712826110' at diff -4931 '-4320'␊
failed at 'l' with '1712826110'␊
failed at 'h' with '1712830997' at diff -4887 '-4320'␊
failed at 'l' with '1712835930' at diff -4933 '-4320'␊
failed at 'l' with '1712835930'␊
failed at 'h' with '1712840817' at diff -4887 '-4320'␊
failed at 'l' with '1712845750' at diff -4933 '-4320'␊
failed at 'l' with '1712845750'␊
  0 1712766950      0 (0us)␊
  1 1712772009  -5059 (-7us)␊
  2 1712776946  -4937 (-6us)␊
  3 1712781833  -4887 (-6us)␊
  4 1712786762  -4929 (-6us)␊
  5 1712791649  -4887 (-6us)␊
  6 1712796582  -4933 (-6us)␊
  7 1712801539  -4957 (-6us)␊
  8 1712806402  -4863 (-5us)␊
  9 1712811359  -4957 (-6us)␊
 10 1712816222  -4863 (-5us)␊
 11 1712821179  -4957 (-6us)␊
 12 1712826110  -4931 (-6us)␊
 13 1712830997  -4887 (-6us)␊
 14 1712835930  -4933 (-6us)␊
 15 1712840817  -4887 (-6us)␊
 16 1712845750  -4933 (-6us)␊
␊
  0 1712766950      0 (0us)␊
␊
failed at 'l' with '1713077036'␊
failed at 'h' with '1713077209'␊
failed at 'l' with '1713087144' at diff -9935 '-9120'␊
failed at 'l' with '1713087144'␊
failed at 'h' with '1713097039' at diff -9895 '-9120'␊
failed at 'l' with '1713106932' at diff -9893 '-9120'␊
failed at 'l' with '1713106932'␊
failed at 'h' with '1713116829' at diff -9897 '-9120'␊
failed at 'l' with '1713126792' at diff -9963 '-9120'␊
failed at 'l' with '1713126792'␊
failed at 'h' with '1713136691' at diff -9899 '-9120'␊
failed at 'l' with '1713146584' at diff -9893 '-9120'␊
failed at 'l' with '1713146584'␊
failed at 'l' with '1713077036'␊
failed at 'h' with '1713082101'␊
failed at 'l' with '1713087050' at diff -4949 '-4320'␊
failed at 'l' with '1713087050'␊
failed at 'h' with '1713092013' at diff -4963 '-4320'␊
failed at 'l' with '1713096942' at diff -4929 '-4320'␊
failed at 'l' with '1713096942'␊
failed at 'h' with '1713101975' at diff -5033 '-4320'␊
failed at 'l' with '1713106838' at diff -4863 '-4320'␊
failed at 'l' with '1713106838'␊
failed at 'h' with '1713111871' at diff -5033 '-4320'␊
failed at 'l' with '1713116734' at diff -4863 '-4320'␊
failed at 'l' with '1713116734'␊
failed at 'h' with '1713121767' at diff -5033 '-4320'␊
failed at 'l' with '1713126698' at diff -4931 '-4320'␊
failed at 'l' with '1713126698'␊
failed at 'h' with '1713131661' at diff -4963 '-4320'␊
failed at 'l' with '1713136594' at diff -4933 '-4320'␊
failed at 'l' with '1713136594'␊
failed at 'h' with '1713141559' at diff -4965 '-4320'␊
failed at 'l' with '1713146490' at diff -4931 '-4320'␊
failed at 'l' with '1713146490'␊
failed at 'h' with '1713151523' at diff -5033 '-4320'␊
failed at 'l' with '1713156386' at diff -4863 '-4320'␊
failed at 'l' with '1713156386'␊
  0 1713077036      0 (0us)␊
  1 1713077209   -173 (-1us)␊
  2 1713087144  -9935 (-14us)␊
  3 1713097039  -9895 (-13us)␊
  4 1713106932  -9893 (-13us)␊
  5 1713116829  -9897 (-13us)␊
  6 1713126792  -9963 (-14us)␊
  7 1713136691  -9899 (-13us)␊
  8 1713146584  -9893 (-13us)␊
␊
failed at 'l' with '1728471712'␊
failed at 'h' with '1728501371' at diff -29659 '-28320'␊
failed at 'l' with '1728540800' at diff -39429 '-28320'␊
failed at 'l' with '1728540800'␊
failed at 'l' with '1728471712'␊
failed at 'h' with '1728476689'␊
failed at 'l' with '1728481630' at diff -4941 '-4320'␊
failed at 'l' with '1728481630'␊
failed at 'h' with '1728486517' at diff -4887 '-4320'␊
failed at 'l' with '1728491446' at diff -4929 '-4320'␊
failed at 'l' with '1728491446'␊
failed at 'h' with '1728496333' at diff -4887 '-4320'␊
failed at 'l' with '1728501266' at diff -4933 '-4320'␊
failed at 'l' with '1728501266'␊
failed at 'h' with '1728506315' at diff -5049 '-4320'␊
failed at 'l' with '1728511178' at diff -4863 '-4320'␊
failed at 'l' with '1728511178'␊
failed at 'h' with '1728516135' at diff -4957 '-4320'␊
failed at 'l' with '1728520998' at diff -4863 '-4320'␊
failed at 'l' with '1728520998'␊
failed at 'h' with '1728525955' at diff -4957 '-4320'␊
failed at 'l' with '1728530886' at diff -4931 '-4320'␊
failed at 'l' with '1728530886'␊
failed at 'h' with '1728535773' at diff -4887 '-4320'␊
failed at 'l' with '1728540706' at diff -4933 '-4320'␊
failed at 'l' with '1728540706'␊
failed at 'h' with '1728545669' at diff -4963 '-4320'␊
failed at 'l' with '1728550602' at diff -4933 '-4320'␊
failed at 'l' with '1728550602'␊
  0 1728471712      0 (0us)␊
  1 1728501371 -29659 (4us)␊
  2 1728540800 -39429 (-8us)␊
␊
  0 1740080998      0 (0us)␊
  1 1740091021 -10023 (-14us)␊
  2 1740130352 -39331 (-7us)␊
␊
failed at 'l' with '1740080998'␊
failed at 'h' with '1740091021' at diff -10023 '-9120'␊
failed at 'l' with '1740130352' at diff -39331 '-9120'␊
failed at 'l' with '1740130352'␊
failed at 'l' with '1740080998'␊
failed at 'h' with '1740085971'␊
failed at 'l' with '1740090918' at diff -4947 '-4320'␊
failed at 'l' with '1740090918'␊
failed at 'h' with '1740095893' at diff -4975 '-4320'␊
failed at 'l' with '1740100822' at diff -4929 '-4320'␊
failed at 'l' with '1740100822'␊
failed at 'h' with '1740105705' at diff -4883 '-4320'␊
failed at 'l' with '1740110634' at diff -4929 '-4320'␊
failed at 'l' with '1740110634'␊
failed at 'h' with '1740115587' at diff -4953 '-4320'␊
failed at 'l' with '1740120450' at diff -4863 '-4320'␊
failed at 'l' with '1740120450'␊
failed at 'h' with '1740125399' at diff -4949 '-4320'␊
failed at 'l' with '1740130262' at diff -4863 '-4320'␊
failed at 'l' with '1740130262'␊
failed at 'h' with '1740135287' at diff -5025 '-4320'␊
failed at 'l' with '1740140218' at diff -4931 '-4320'␊
failed at 'l' with '1740140218'␊
failed at 'h' with '1740145101' at diff -4883 '-4320'␊
failed at 'l' with '1740150030' at diff -4929 '-4320'␊
failed at 'l' with '1740150030'␊
failed at 'h' with '1740154913' at diff -4883 '-4320'␊
failed at 'l' with '1740159842' at diff -4929 '-4320'␊
failed at 'l' with '1740159842'␊
>>> spi_device_test␊
>>> time_test␊
>>> timeout_test␊
>>> timestamp_test␊
>>> tmp12x_test␊
>>> vector1_test␊
>>> vector2_test␊
>>> vector3_test␊
>>> vector4_test␊
>>> vector_test␊
␊
Failed 11 of 6026 tests␊
```

</details>


### Unittests Nucleo L552

<details>
<summary>Passed 4876 tests</summary>

```
Unittests (Apr  2 2022, 16:32:04)␊
Machine:  rleh-pc␊
User:     rleh␊
Os:       Linux-5.16.16-200.fc35.x86_64-x86_64-with-glibc2.34␊
Compiler: arm-none-eabi-g++ 10.2.1␊
Local Git User:␊
Name:  Raphael Lehmann␊
Email: raphael@rleh.de␊
Last Commit:␊
SHA:             175f915e272a1663025e36a515d3f135353cceae␊
Abbreviated SHA: 175f915e27␊
Subject:         [docs] Add Christopher as maintainer␊
␊
Author:␊
Name:      Christopher Durand␊
Email:     christopher.durand@rwth-aachen.de␊
Date:      2022-04-01 22:37:05+02:00␊
Timestamp: 1648845425␊
␊
Committer:␊
Name:      Christopher Durand␊
Email:     christopher.durand@rwth-aachen.de␊
Date:      2022-04-01 22:37:05+02:00␊
Timestamp: 1648845425␊
␊
File Status:␊
Modified:  0␊
Added:     0␊
Deleted:   0␊
Renamed:   0␊
Copied:    0␊
Untracked: 0␊
␊
>>> LUDecomposition_test␊
>>> ad7280a_test␊
>>> angle_test␊
>>> arithmetic_traits_test␊
>>> atomic_queue_test␊
>>> bit_operation_test␊
>>> block_allocator_test␊
>>> bme280_test␊
>>> bmp085_test␊
>>> bounded_deque_test␊
>>> bounded_queue_test␊
>>> bounded_stack_test␊
>>> button_group_test␊
>>> button_test␊
>>> can_bit_timings_test␊
>>> can_connector_base_test␊
>>> can_connector_test␊
>>> can_lawicel_formatter_test␊
>>> can_message_test␊
>>> circle_2d_test␊
>>> clock_test␊
>>> color_test␊
>>> debounce_test␊
>>> delay_test␊
>>> dispatcher_test␊
>>> doubly_linked_list_test␊
>>> drv832x_spi_test␊
>>> dynamic_array_test␊
>>> endianness_test␊
>>> fiber_test␊
>>> fir_test␊
>>> flash_test␊
>>> i2c_test␊
>>> inplace_any_test␊
>>> interface_test␊
>>> interface_test␊
>>> io_stream_test␊
>>> lagrange_interpolation_test␊
>>> line_2d_test␊
>>> line_segment_2d_test␊
>>> linear_interpolation_test␊
>>> linked_list_test␊
>>> location_2d_test␊
>>> ltc2984_test␊
>>> matrix_test␊
>>> matrix_vector_test␊
>>> mcp2515_can_bit_timings_test␊
>>> median_test␊
>>> message_test␊
>>> moving_average_test␊
>>> node_test␊
>>> operator_test␊
>>> pair_test␊
>>> periodic_timer_test␊
>>> pid_test␊
>>> point_set_2d_test␊
>>> polygon_2d_test␊
>>> prescaler_test␊
>>> protothread_test␊
>>> quaternion_test␊
>>> ramp_test␊
>>> range_test␊
>>> ray_2d_test␊
>>> register_test␊
>>> resumable_test␊
>>> s_curve_controller_test␊
>>> s_curve_generator_test␊
>>> saturation_test␊
>>> scheduler_test␊
>>> slave_test␊
>>> spi_device_test␊
>>> time_test␊
>>> timeout_test␊
>>> timestamp_test␊
>>> tmp12x_test␊
>>> vector1_test␊
>>> vector2_test␊
>>> vector3_test␊
>>> vector4_test␊
>>> vector_test␊
␊
Passed 4876 tests␊
OK!
```

</details>

### Unittests Nucleo F091

<details>
<summary>Assertion 'nvic.undef' @ 0xFFFFFFF3 (4294967283) failed!␊ Abandoning...␊</summary>

```
Unittests (Apr  2 2022, 17:02:30)␊
Machine:  rleh-pc␊
User:     rleh␊
Os:       Linux-5.16.16-200.fc35.x86_64-x86_64-with-glibc2.34␊
Compiler: arm-none-eabi-g++ 10.2.1␊
Local Git User:␊
Name:  Raphael Lehmann␊
Email: raphael@rleh.de␊
Last Commit:␊
SHA:             175f915e272a1663025e36a515d3f135353cceae␊
Abbreviated SHA: 175f915e27␊
Subject:         [docs] Add Christopher as maintainer␊
␊
Author:␊
Name:      Christopher Durand␊
Email:     christopher.durand@rwth-aachen.de␊
Date:      2022-04-01 22:37:05+02:00␊
Timestamp: 1648845425␊
␊
Committer:␊
Name:      Christopher Durand␊
Email:     christopher.durand@rwth-aachen.de␊
Date:      2022-04-01 22:37:05+02:00␊
Timestamp: 1648845425␊
␊
File Status:␊
Modified:  0␊
Added:     0␊
Deleted:   0␊
Renamed:   0␊
Copied:    0␊
Untracked: 0␊
␊
>>> LUDecomposition_test␊
>>> ad7280a_test␊
>>> angle_test␊
>>> arithmetic_traits_test␊
>>> atomic_queue_test␊
>>> bit_operation_test␊
>>> block_allocator_test␊
>>> bme280_test␊
>>> bmp085_test␊
>>> bounded_deque_test␊
>>> bounded_queue_test␊
>>> bounded_stack_test␊
>>> button_group_test␊
>>> button_test␊
>>> can_bit_timings_test␊
>>> can_connector_base_test␊
>>> can_connector_test␊
>>> can_lawicel_formatter_test␊
>>> can_message_test␊
>>> circle_2d_test␊
>>> clock_test␊
>>> color_test␊
>>> debounce_test␊
>>> delay_test␊
>>> dispatcher_test␊
>>> doubly_linked_list_test␊
>>> drv832x_spi_test␊
>>> dynamic_array_test␊
>>> endianness_test␊
>>> fiber_test␊
>>> fir_test␊
>>> flash_test␊
>>> i2c_test␊
>>> inplace_any_test␊
>>> interface_test␊
>>> interface_test␊
>>> io_stream_test␊
>>> lagrange_interpolation_test␊
>>> line_2d_test␊
>>> line_segment_2d_test␊
>>> linear_interpolation_test␊
>>> linked_list_test␊
>>> location_2d_test␊
>>> ltc2984_test␊
>>> matrix_test␊
>>> matrix_vector_test␊
>>> mcp2515_can_bit_timings_test␊
>>> median_test␊
>>> message_test␊
>>> moving_average_test␊
>>> node_test␊
>>> operator_test␊
>>> pair_test␊
>>> periodic_timer_test␊
>>> pid_test␊
>>> point_set_2d_test␊
>>> polygon_2d_test␊
>>> prescaler_test␊
>>> protothread_test␊
>>> quaternion_test␊
>>> ramp_test␊
>>> range_test␊
>>> ray_2d_test␊
>>> register_test␊
>>> resumable_test␊
>>> s_curve_controller_test␊
>>> s_curve_generator_test␊
>>> saturation_test␊
>>> scheduler_test␊
>>> slave_test␊
Assertion 'nvic.undef' @ 0xFFFFFFF3 (4294967283) failed!␊
Abandoning...␊
```

</details>

rleh commented:
Only the gpio sampler/test and spi bitbang unittests are failing, same as last release (https://github.com/modm-io/modm/pull/797#issuecomment-1003432334).
And Nucleo-F091 still run out of memory...

rleh commented:
@salkinium Could you check the unittests on MacOS?

rleh commented on docs/release/2022q1.md:
```diff
@@ -0,0 +1,141 @@
+- Raspberry Pi PR2040 Support.
+- Microchip SAM Timer driver.
+- Improved Project and Doxygen documentation.
+- STM32G4 internal flash driver.
```
```suggestion
- STM32G4 internal flash driver.
- TinyUSB on STM32G4.
```

rleh reviewed (commented):
Approval from me ✅

chris-durand commented on docs/release/2022q1.md:
```diff
@@ -0,0 +1,142 @@
+
+Features:
+
+- STM32L5 Support.
+- Raspberry Pi PR2040 Support.
+- Microchip SAM Timer driver.
+- Improved Project and Doxygen documentation.
```
```suggestion
- STM32L5 support.
- Raspberry Pi PR2040 support.
- Microchip SAM timer driver.
- Improved project and Doxygen documentation.
```
Minor spelling nitpick, common nouns like "support" and "project" should not be capitalized here.

chris-durand commented on docs/release/2022q1.md:
```diff
@@ -0,0 +1,142 @@
+## 2022-04-01: 2022q1 release
+
+This release covers everything from 2022-01-01 and has been tested with avr-gcc
+v11.2.0 from Upstream and arm-none-eabi-gcc 10.3-2021.10 from Arm.
```
```suggestion
v11.2.0 from upstream and arm-none-eabi-gcc 10.3-2021.10 from Arm.
```

salkinium commented:
Thanks for doing the release! Nice to know the modm bus factor is not 1 anymore ;-P

https://doc.modm.io needs to get the 2022q1 release, I guess by copying a folder somewhere?

chris-durand commented:
> https://doc.modm.io needs to get the 2022q1 release, I guess by copying a folder somewhere?

We should automate that for the next release with an [on release](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#release) github action.

rleh commented:
> https://doc.modm.io needs to get the 2022q1 release, I guess by copying a folder somewhere?

Yes, already done (5 minutes ago). Thanks for reminding.

> We should automate that for the next release with an [on release](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#release) github action.

Won't work because the CI has very limited access to the server that hosts https://docs.modm.io (for security reasons).

chris-durand commented:
> Won't work because the CI has very limited access to the server that hosts https://docs.modm.io (for security reasons).

I don't see what the difference is to uploading the `develop` documentation. The current doc deployment job does the following:

```
- name: Upload api documentation to docs.modm.io
  if: always()
  env:
    DOCS_MODM_IO_UPLOAD: ${{ secrets.DOCS_MODM_IO_UPLOAD }}
  run: |
    curl https://docs.modm.io/upload/compressed --upload-file modm-api-docs.tar.gz --user $DOCS_MODM_IO_UPLOAD
```

The built file is uploaded via a https request and something on the server extracts the archive to the right place. A simplistic solution for doing it for the release docs could be using a different file name that contains the release name (e.g. `modm-api-docs-release-2022q2.tar.gz`) or a url parameter and the correct extraction is handled on the server side. I don't see any new security problems with such an approach compared to the `develop` documentation.

rleh commented:
Actually, we have a problem with https://docs.modm.io hosting: The filesystem (ext4) dedicated for docs.modm.io has run out of inodes:

```bash
$ df -i
Filesystem      Inodes   IUsed  IFree IUse% Mounted on
[...]
/dev/vdaX      2692224 2692224      0  100% /mnt/docs-modm-io-fs
```

because a single release adds nearly 1 million inodes/files:

```bash
$ find 2021q4/ | wc -l
788108
```

salkinium commented:
Let's reformat the entire disk to add a new filesystem for every release!1!!

rleh commented:
> Let's reformat the entire disk to add a new filesystem for every release!1!!

Done. 7882287 inodes are left for future releases 😆
